### PR TITLE
[Audio] SSL Pretraining framework for flow-matching model for audio processing

### PR DESCRIPTION
**Pull request by Kuray107**
# What does this PR do ?

This PR adds a ssl-pretraining framework and a flow matching AudioToAudio model

**Collection**: audio

# Changelog 
- Add a flow matching generative model in `audio/models/enhancement.py`
- Create `audio/modules/ssl_pretrain_masking` to partially-masked out input in the ssl pre-training stage.
- Create `audio/parts/submodules/transformerunet.py` to implement a Transformer U-Net structure.
- Create `audio/parts/submodules/flow.py` for training/inferencing with the flow matching generative model
- Create `audio/parts/utils/callbacks.py` to add processed audio in the training logger.

# Usage

Model can be first be pre-trained using

```python
python examples/audio/audio_to_audio_train.py --config-name flow_matching_generative_ssl_pretraining.yaml
```
- Please update the `train_ds.manifest_filepath` and `validation_ds.manifest_filepath` with the correct filepaths

After the ssl-pretraining stage is completed, the pretrained model could be finetuned using 

```python
python examples/audio/audio_to_audio_train.py --config-name flow_matching_generative_finetuning.yaml
```
- Remember to update the `init_from_nemo_model` with the pretrained checkpoint path in the config file.

Alternatively, one can also just train a flow matching generative from scratch:
```python
python examples/audio/audio_to_audio_train.py --config-name flow_matching_generative.yaml
```

Model can be evaluated using
```python
python examples/audio/audio_to_audio_eval.py \
    model_path=${PATH_TO_NEMO_FILE} \
    dataset_manifest=${PATH_TO_DATASET_MANIFEST} \
    output_dir=${PATH_TO_OUTPUT_DIR} \
    input_key=noisy_filepath \
    target_key=clean_filepath \
    metrics=[sdr,estoi]
```

Sampler setting can be easily overridden, e.g., to change the number of sampling steps as

```python
python examples/audio/audio_to_audio_eval.py ... ++sampler.num_steps=${NUM_STEPS}
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: nemo/collections/audio/parts/submodules/transformerunet.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+
+class GEGLU(Module):
+    """The GeGLU activation implementation"""
+
+    def forward(self, x: torch.Tensor):
+        x, gate = x.chunk(2, dim=-1)
+        return F.gelu(gate) * x
+
+
+def FeedForward(dim: int, mult: int = 4, dropout: float = 0.0):
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/transformerunet.py:182]: Fixed by changing the name to `get_feedforward_layer`

File: examples/audio/conf/flow_matching_generative.yaml [hunk 2]
```diff
@@ -0,0 +1,164 @@
+name: flow_matching_generative
+
+model:
+  type: flow_matching
+  sample_rate: 16000
+  skip_nan_grad: false
+  num_outputs: 1
+  p_cond: 0.9 # Proability of feeding the conditional input into the model.
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply examples/audio/conf/flow_matching_generative.yaml:8]: `p_cond` is only effective during the training process. In the inference stage one could choose whether to use classifier-free guidance or not.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        Return the mean of p_t(x | start_state, end_state) at time t
+        """
+        pass
+
+    @abstractmethod
+    def std(self, *, time: torch.Tensor, start_state: torch.Tensor, end_state: torch.Tensor) -> torch.Tensor:
+        """
+        Return the standard deviation of p_t(x | start_state, end_state) at time t
+        """
+        pass
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:54]: Good point. I'll change `start_state` to `x_start` and `end_state` to `x_end` through the PR.

[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:54]: Fixed

File: nemo/collections/audio/parts/submodules/flow.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+    def sample(self, *, time: torch.Tensor, start_state: torch.Tensor, end_state: torch.Tensor) -> torch.Tensor:
+        """
+        Generate a sample from p_t(x | start_state, end_state) at time t
+        """
+        time = self._broadcast_time(time, n_dim=start_state.ndim)
+
+        mean = self.mean(time=time, start_state=start_state, end_state=end_state)
+        std = self.std(time=time, start_state=start_state, end_state=end_state)
+        return mean + std * torch.randn_like(mean)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:95]: Fixed by adding the above comment.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+    def flow(
+        self, *, time: torch.Tensor, start_state: torch.Tensor, end_state: torch.Tensor, point: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute the conditional flow phi_t( point | start_state, end_state)
+        """
+        mean = self.mean(time=time, start_state=start_state, end_state=end_state)
+        std = self.std(time=time, start_state=start_state, end_state=end_state)
+        return mean + std * (point - start_state)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:122]: Fixed by adding the comment.

File: nemo/collections/audio/models/enhancement.py [hunk 6]
```diff
@@ -618,6 +619,265 @@ def evaluation_step(self, batch, batch_idx, dataloader_idx: int = 0, tag: str =
...
+        # Self-Supervised Pretraining
+        if self.ssl_pretrain_masking is not None:
+            input_enc = self.ssl_pretrain_masking(input_spec=input_enc, length=input_enc_len)
+
+        # Drop off conditional inputs (input_enc) with (1 - p_cond) probability.
+        # The dropped conditions will be set to zeros
+        keep_conditions = einops.rearrange((torch.rand(batch_size) < self.p_cond).float(), 'B -> B 1 1 1')
+        input_enc = input_enc * keep_conditions.to(input_enc.device)
+
+        start_state = torch.zeros_like(input_enc)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/models/enhancement.py:775]: Fixed

File: nemo/collections/audio/modules/ssl_pretrain_masking.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            if min_len < self.patch_size * mask_patches:
+                mask_patches = min_len // self.patch_size
+
+            for idx in range(input_spec.shape[0]):
+                cur_len = length[idx]
+                patches = range(cur_len // self.patch_size)
+                masked_patches = random.sample(patches, mask_patches)
+                for mp in masked_patches:
+                    augmented_spec[idx, :, :, mp * self.patch_size : (mp + 1) * self.patch_size] = 0.0
+        else:
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:81]: This is for the validation purpose. In the pretraining stage we still would like to have a way to evaluate the checkpoints and avoid overfitting. In the validation stage (`self.training == False`) we fix the masking pattern instead of generating it randomly.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+            vector_field, _ = self.estimator(input=estimator_input, input_length=state_length, condition=time)
+
+            state = state + vector_field * self.time_step
+
+            if state_length is not None:
+                state = mask_sequence_tensor(state, state_length)
+
+        if state_length is not None:
+            state = mask_sequence_tensor(state, state_length)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:291]: Fixed by removing it.

File: nemo/collections/audio/parts/submodules/transformerunet.py [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+
+__all__ = ['TransformerUNet']
+
+
+def exists(val):
+    return val is not None
+
+
+def default(val, d):
+    return val if exists(val) else d
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/transformerunet.py:59]: Fixed by inlining all `default`.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    """
+    Abstract class for different conditional flow-matching (CFM) classes
+
+    Time horizon is [time_min, time_max (should be 1)]
+
+    every path is "conditioned" on endpoints of the path
+    endpoints are just our paired data samples
+    subclasses need to either:
+       1. implement mean, std, d_mean, d_std
+       2. implement mean, std; override vector_field, flow (also d_flow if needed)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:34]: Fixed as suggested. I also removed the generic implementation of the vector field and change it to another abstract method.

File: nemo/collections/audio/modules/ssl_pretrain_masking.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+class SSLPretrainWithMaskedPatch(NeuralModule):
+    """
+    Zeroes out fixed size time patches of the spectrogram.
+    All samples in batch are guaranteed to have the same amount of masked time steps.
+
+    Args:
+        patch_size (int): up to how many time steps does one patch consist of.
+            Defaults to 48.
+        mask_fraction (float): how much fraction in each sample to be masked (number of patches is rounded up).
+            Range from 0.0 to 1.0. Defaults to 0.7.
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:35]: This code is originally from `nemo.collections.asr.modules.audio_preprocessing.MaskedPatchAugmentation`. I removed some unnecessary arguments but they have the same behavior to perform the time-frame masking. 

As for the example in the comment, the only possible mask is to mask out 0~47 frames. 

Do you suggest we make the arguments clearer, or do you think this masking implementation is confusing and should be modified?

File: nemo/collections/audio/modules/ssl_pretrain_masking.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    def forward(self, input_spec, length):
+        augmented_spec = input_spec
+
+        min_len = torch.min(length)
+        if self.training:
+            len_fraction = int(min_len * self.mask_fraction)
+            mask_patches = len_fraction // self.patch_size + int(len_fraction % self.patch_size != 0)
+
+            if min_len < self.patch_size * mask_patches:
+                mask_patches = min_len // self.patch_size
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:73]: Let's say `mask_fraction=0.7` and `patch_size=10`. Since `min_len = 87`, we'll obtain `mask_patches=7`. Each of the two data will have 7 patches of 10-frame mask. Therefore, for unbalanced batch the masked fraction will be impacted. 

For the pretraining on libri-light this won't be a problem, as most (if not all) data are longer than our training data length setting. However, this could be problematic when training on other datasets with very unbalanced data.

[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:73]: @anteju  @racoiaws  What do you suggest on this implementaiton?

File: nemo/collections/audio/modules/ssl_pretrain_masking.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+            for idx in range(input_spec.shape[0]):
+                cur_len = length[idx]
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:76]: Fixed as suggested.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+        return (self.time_max - time) * x_start + time * x_end
+
+    def std(self, *, x_start: torch.Tensor, x_end: torch.Tensor, time: torch.Tensor) -> torch.Tensor:
+        return (self.time_max - time) * self.sigma_start + time * self.sigma_end
+
+    def d_mean(self, *, x_start: torch.Tensor, x_end: torch.Tensor, time: torch.Tensor) -> torch.Tensor:
+        return x_end - x_start
+
+    def d_std(self, *, x_start: torch.Tensor, x_end: torch.Tensor, time: torch.Tensor) -> torch.Tensor:
+        return self.sigma_end - self.sigma_start
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:154]: Fixed.

File: nemo/collections/audio/parts/submodules/flow.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+    ) -> torch.Tensor:
+        """
+        Compute the conditional vector field v_t( point | x_start, x_end)
+        """
+        pass
+
+    @staticmethod
+    def _broadcast_time(time: torch.Tensor, n_dim: int) -> torch.Tensor:
+        """
+        Broadcast time tensor to the desired number of dimension
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:68]: Fixed.

File: nemo/collections/audio/parts/submodules/flow.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/parts/submodules/flow.py:34]: Fixed.

File: nemo/collections/audio/modules/ssl_pretrain_masking.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/modules/ssl_pretrain_masking.py:73]: Understood. I added the comment and example in `forward()` docstring.

File: examples/audio/audio_to_audio_train.py [hunk 16]
```diff
@@ -50,6 +51,7 @@ class ModelType(str, Enum):
     Predictive = 'predictive'
     ScoreBased = 'score_based'
     SchroedingerBridge = 'schroedinger_bridge'
+    FlowMatchingAudioToAudioModel = 'flow_matching'
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply examples/audio/audio_to_audio_train.py:54]: Fixed

File: examples/audio/audio_to_audio_train.py [hunk 17]
```diff
@@ -62,6 +64,8 @@ def get_model_class(model_type: ModelType):
         return ScoreBasedGenerativeAudioToAudioModel
     elif model_type == ModelType.SchroedingerBridge:
         return SchroedingerBridgeAudioToAudioModel
+    elif model_type == ModelType.FlowMatchingAudioToAudioModel:
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply examples/audio/audio_to_audio_train.py:67]: Fixed

File: nemo/collections/audio/data/audio_to_audio_lhotse.py [hunk 18]
```diff
@@ -44,19 +44,29 @@ class LhotseAudioToTargetDataset(torch.utils.data.Dataset):
     EMBEDDING_KEY = "embedding_vector"
 
     def __getitem__(self, cuts: CutSet) -> dict[str, torch.Tensor]:
-        src_audio, src_audio_lens = collate_audio(cuts)
+        # In the rare case, the collate_audio function would raise the FileSeek error when loading .flac (https://github.com/bastibe/python-soundfile/issues/274)
+        # A workaround is to use fault_tolerant and skip failed data, resulting to a smaller batch size for the few problematic cases.
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/data/audio_to_audio_lhotse.py:48]: Fixed

File: nemo/collections/audio/models/enhancement.py [hunk 19]
```diff
@@ -618,6 +619,265 @@ def evaluation_step(self, batch, batch_idx, dataloader_idx: int = 0, tag: str =
         return {f'{tag}_loss': loss}
 
 
+class FlowMatchingAudioToAudioModel(AudioToAudioModel):
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/models/enhancement.py:622]: Done

File: nemo/collections/audio/models/enhancement.py [hunk 20]
```diff
@@ -618,6 +619,265 @@ def evaluation_step(self, batch, batch_idx, dataloader_idx: int = 0, tag: str =
...
+            input_signal = input_signal / (norm_scale + self.eps)
+
+        # Encoder
+        encoded, encoded_length = self.encoder(input=input_signal, input_length=input_length)
+
+        if self.p_cond == 0:
+            encoded = torch.zeros_like(encoded)
+
+        if self.ssl_pretrain_masking is not None:
+            encoded = self.ssl_pretrain_masking(input_spec=encoded, length=encoded_length)
```
[the remark this answers is not in the document]
Comment (Kuray107) [reply nemo/collections/audio/models/enhancement.py:720]: Fixed.
